### PR TITLE
Add notes on any source and any port

### PR DESCRIPTION
**Pull request by susam**
See https://github.com/cloudmarker/guides/blob/any-source-any-port/tech/firewall-rule-format.md for the rendered README.

Here are the direct links to the rendered new sections:

- [Any Source](https://github.com/cloudmarker/guides/blob/any-source-any-port/tech/firewall-rule-format.md#any-source)
- [Any Port](https://github.com/cloudmarker/guides/blob/any-source-any-port/tech/firewall-rule-format.md#any-port)

### CODE REVIEW

File: tech/firewall-rule-format.md [hunk 1]
```diff
@@ -489,7 +514,7 @@ potentially weak:
     if they are allowed to all source address ranges.
```
Comment (s-nirali) [inline tech/firewall-rule-format.md:528]: I missed this in my earlier review. In the [Multiple Ranges](https://github.com/cloudmarker/guides/blob/any-source-any-port/tech/firewall-rule-format.md#multiple-ranges) section, the sentence "There are primarily three important pieces of information..." should be "There are primarily **four** important pieces of information...".

File: tech/firewall-rule-format.md [hunk 2]
```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
...
+      "direction": "Inbound",
+      "provisioning_state": "Succeeded",
+      "name": "Port_22",
+      "etag": "W/\"a9ceecf4-eb4a-4fb8-b961-4cd9ab246e70\""
+    }
+
+
+### Any Source: GCP
+
+GCP contains this firewall rule to allow connectivity port 22 from
```
Comment (s-nirali) [inline tech/firewall-rule-format.md]: "GCP contains this firewall rule to allow connectivity **to** port 22 from"

File: tech/firewall-rule-format.md [hunk 3]
```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
...
+      "kind": "compute#firewall",
+      "id": "4387657442666205482",
+      "creationTimestamp": "2019-03-26T11:30:29.863-07:00",
+      "name": "allow-8000",
+      "description": "",
+      "network": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/networks/default",
+      "priority": 1000,
+      "sourceRanges": [
+        "0.0.0.0/0"
+      ],
```
Comment (s-nirali) [inline tech/firewall-rule-format.md:824]: I was just wondering, is the case for multiple CIDRs not applicable for GCP?

File: tech/firewall-rule-format.md [hunk 4]
```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
...
+
+In this section, we will see a couple of ways we can allow connections
+to all ports.
+
+
+#### `destination_port_range` set to `*`
+
+Here is an example where an inbound security rule is created with
+*Destination port ranges* set to `*`.
+
```
Comment (s-nirali) [inline tech/firewall-rule-format.md:880]: `Destination port range` is "*". `Destination port ranges` is "[]".

File: tech/firewall-rule-format.md [hunk 5]
```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
...
+      "provisioning_state": "Succeeded",
+      "name": "Port_All",
+      "etag": "W/\"2873cc0e-f41e-4c94-a773-8b0982d0db55\""
+    }
+
+
+#### `destination_port_range` set to integer range
+
+Another way to allow access to all ports is to set *Destination port
+ranges* to `0-65535`.
```
Comment (s-nirali) [inline tech/firewall-rule-format.md:904]: The key here is also `destination_port_range`.

**Review by s-nirali — APPROVED** [review]

File: tech/firewall-rule-format.md [hunk 6]
```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
...
+
+
+#### `source_address_prefix` set to CIDR
+
+Azure portal did not allow us to configure a rule with source address
+range as `0.0.0.0/0`. An attempt to do so resulted in this error:
+`Invalid argument: 'Prefix'. Reason: The prefix must be between 1 and 32.`.
+
+The weakest rule we could configure in Azure portal using a single CIDR
+is `0.0.0.0/1`. Whe an inbound security rule is configured with *IP
```
Comment (s-nirali) [inline tech/firewall-rule-format.md]: `When an..`

File: tech/firewall-rule-format.md [hunk 7]
```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
...
+      "direction": "Inbound",
+      "provisioning_state": "Succeeded",
+      "name": "Port_All",
+      "etag": "W/\"7f844086-b511-4fbb-b14e-67bb6d1635a6\""
+    }
+
+
+An interesting thing worth figuring out was if we could create a rule
+with *Destination port ranges* as something like `*,8080` to know if we
+we need to handle formats like these as well. An attempt to create a
```
Comment (s-nirali) [inline tech/firewall-rule-format.md]: Single `we`.

File: tech/firewall-rule-format.md (hunk 1 above)
Comment (susam) [reply tech/firewall-rule-format.md:528]: Thank you for pointing this out. I will fix this in a separate pull request in order to not mix separate concerns in the same pull request. In fact, based on cloudmarker/cloudmarker#69 that I am working on, this section needs a few more updates too which I will take care of in a separate pull request.

File: tech/firewall-rule-format.md (hunk 3 above)
Comment (susam) [reply tech/firewall-rule-format.md:824]: It is still applicable. The reason why we have `"source_address_prefixes": ["0.0.0.0/1", "128.0.0.0/1"]` in the Azure example but we only have `"sourceRanges": ["0.0.0.0/0"]` in the GCP example is that the Azure portal did not let me enter `0.0.0.0/0` as a valid CIDR at all, so just for the sake of writing an example I entered two CIDRs that when combined is equivalent to `0.0.0.0/0`.

Having said that, I am not planning to perform union of CIDRs to see if it is equivalent to the entire Internet in the initial implementation. That is a very interesting problem to solve though.

File: tech/firewall-rule-format.md (hunk 4 above)
Comment (susam) [reply tech/firewall-rule-format.md:880]: The italicized text are labels I have picked up verbatim from Azure Portal UI. So here I mean that I entered `*` in the *Destination port ranges* in the Azure Portal UI.

Screen Shot 2019-04-01 at 2 44 02 PM

Azure decides to put the port range in either `destination_port_range` or `destination_port_ranges` in the API response depending on whether one or more than one port ranges are specified.

Does this make sense now?

I will clarify this notation convention somewhere at the top of the document and update this pull request.

File: tech/firewall-rule-format.md (hunk 5 above)
Comment (susam) [reply tech/firewall-rule-format.md:904]: See https://github.com/cloudmarker/guides/pull/5#discussion_r270779322.

File: tech/firewall-rule-format.md (hunk 2 above)
Comment (susam) [reply tech/firewall-rule-format.md]: Addressed in [line 781](https://github.com/cloudmarker/guides/pull/5/files#diff-112a678346155cbc8647b65f47882715R781).

File: tech/firewall-rule-format.md (hunk 4 above)
Comment (susam) [reply tech/firewall-rule-format.md:880]: Clarified the notation at [line 46](https://github.com/cloudmarker/guides/pull/5/files#diff-112a678346155cbc8647b65f47882715R46).

File: tech/firewall-rule-format.md (hunk 6 above)
Comment (susam) [reply tech/firewall-rule-format.md]: Fixed at [line 682](https://github.com/cloudmarker/guides/pull/5/files#diff-112a678346155cbc8647b65f47882715R682).

File: tech/firewall-rule-format.md (hunk 7 above)
Comment (susam) [reply tech/firewall-rule-format.md]: Fixed at [line 962](https://github.com/cloudmarker/guides/pull/5/files#diff-112a678346155cbc8647b65f47882715R962).

**Comment by susam** [comment]
Addressed the review comments. The pull request is already in approved state. So I will go ahead and merge this. If there are more review comments, please feel free to share them. I will address them in a future pull request.

File: tech/firewall-rule-format.md (hunk 4 above)
Comment (s-nirali) [reply tech/firewall-rule-format.md:880]: This makes sense. Thank you for the clarification.

**Pull request merged**